README.md
# os3-copier
OpenShift 3.x Version of the [k8s-copier](https://github.com/dweber019/k8s-copier)

This operator aims to provide Kubernetes CRD's to copy resources from one namespace into another.  
This new CRD is called `CopyResrouce` and the resulting resource is called target resource.

## Implemented resource types
- v1.Secret
- v1.ConfigMap

## Usage
You can find some usage examples in `config/samples/**`.

### Configuration
| Name                    | Type    | Default |
| ------------------------|---------|---------|
| WATCH_NAMESPACE         | env var | NA      |
| SYNC_PERIOD             | env var | 300     |
| metrics-addr            | flag    | :8080   |
| enable-leader-election  | flag    | false   |
| dev-mode-enabled        | flag    | false   |

### Permissions
You need a service account to operate your operator. This service account needs to have
access to the target namespace regarding the resource types.  
You can find examples in `config/samples/**`.

### Behavior
If you delete a CopyResource the target resource won't be deleted as it's possible that other implementation depend on it.

## Development setup
### Conventional commits
Execute the following terminal command in the root:
```
curl -o- https://raw.githubusercontent.com/craicoverflow/sailr/master/scripts/install.sh | bash
```
### Go
Follow the installation guide at https://golang.org/doc/install

### Operator-SDK
Uses the [Operator-SDK Version 0.18.2](https://github.com/operator-framework/operator-sdk/blob/v0.18.2) to be able to create and use CRDs compatible with OpenShift 3.x

## Development
For development, you should use `minishift` or any other possible kubernetes compatible implementation, as advised by the operator-sdk framework.

### Update the CRDs
To update the CRDs use the following command
```
make generate
make manifests
```
The files will be generated into `config/**`.

### Build and run
To install the CRDs to the Kubernetes cluster and run the operator outside of Kubernetes use
```
make install run
```

## Deployment
### Automated
The deployment and therefore publishing a Docker image is fully automated with GitHub workflows.

### Manually
Use the automated way over GitHub!!!
To update the docker image use this command
```
make docker-build docker-push IMG=docker.io/baloise/os3-copier:v0.0.1
```
After this you can run with
```
make deploy IMG=docker.io/baloise/os3-copier:v0.0.1
```

## Useful links
- [Kubebilder](https://book.kubebuilder.io)
- [Operator tutorial](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/)
- [Conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)

CHANGELOG.md
## [2.1.0](https://github.com/baloise/os3-copier/compare/v2.0.0...v2.1.0) (2021-10-25)


### Features

* **CHIST-5791:** add targetName feature ([eb6e0eb](https://github.com/baloise/os3-copier/commit/eb6e0ebed171fbdf10a906a827ad9c2eaa78b613))

## [2.0.0](https://github.com/baloise/os3-copier/compare/v1.3.0...v2.0.0) (2021-10-21)


### ⚠ BREAKING CHANGES

* **CHIST-3146:** no

### Features

* **CHIST-3146:** add endpoints for live- and readiness probe ([27c2eb2](https://github.com/baloise/os3-copier/commit/27c2eb2adf81853491e6de540261c3ca1f1f466f))

## [1.3.0](https://github.com/baloise/os3-copier/compare/v1.2.3...v1.3.0) (2021-10-20)


### Features

* **CHIST-3146:** improve error handling and add dev mode ([916055d](https://github.com/baloise/os3-copier/commit/916055d35d1955f443aa0765887a3e32d3127a21))
* **CHIST-3146:** review with Michael ([3eb8050](https://github.com/baloise/os3-copier/commit/3eb8050e24671f91cf45e116e9926e45b55944f8))

### [1.2.3](https://github.com/baloise/os3-copier/compare/v1.2.2...v1.2.3) (2021-10-18)


### Bug Fixes

* **CHIST-1:** fix dublicate domain name in group name, fix role definition, add env var for syncPeriod, ([62d9681](https://github.com/baloise/os3-copier/commit/62d9681ae1ddf71eda455a5dd80fe0af2fd5cc4e))

### [1.2.2](https://github.com/baloise/os3-copier/compare/v1.2.1...v1.2.2) (2021-10-18)


### Bug Fixes

* **CHIST-1:** fix dublicate domain name in group name ([e859a55](https://github.com/baloise/os3-copier/commit/e859a55bcb0ae3740c94af791ca0403ac4237f3d))

### [1.2.1](https://github.com/baloise/os3-copier/compare/v1.2.0...v1.2.1) (2021-10-17)


### Bug Fixes

* **chist-1:** search target resource using an unstructured type ([772ef15](https://github.com/baloise/os3-copier/commit/772ef150e41487b9243c28129c748be0d3a6f4b9))

## [1.2.0](https://github.com/baloise/os3-copier/compare/v1.1.2...v1.2.0) (2021-10-15)


### Features

* **CHIST-1:** add OwnerReference ([9702386](https://github.com/baloise/os3-copier/commit/9702386c5c1af5bb7d388017cbd0e14aaa2f06d7))

### [1.1.2](https://github.com/baloise/os3-copier/compare/v1.1.1...v1.1.2) (2021-10-15)


### Bug Fixes

* **CHIST-1:** fix sync ([d57a01a](https://github.com/baloise/os3-copier/commit/d57a01a6064fbc7b1d7184141bc3a724d42243b8))

### [1.1.1](https://github.com/baloise/os3-copier/compare/v1.1.0...v1.1.1) (2021-10-03)


### Bug Fixes

* **chist-1:** fix resourceVersion validation ([3f14a65](https://github.com/baloise/os3-copier/commit/3f14a650afc5c9e32df2358dcaf3711be2cb1a4d))

## [1.1.0](https://github.com/baloise/os3-copier/compare/v1.0.0...v1.1.0) (2021-10-03)


### Features

* **chist-1:** add health endpoint and log statements ([22baf54](https://github.com/baloise/os3-copier/commit/22baf54c23c4a841492482628ba8598521a1e5d2))

## 1.0.0 (2021-09-30)


### Features

* **CHIST-1:** add changelog ([2b26caa](https://github.com/baloise/os3-copier/commit/2b26caa2f829547b5c0b1c179e4cb2cd5637ba47))
* **CHIST-1:** add VERSION to Makefile ([faaca38](https://github.com/baloise/os3-copier/commit/faaca38afb8bb544e0f09435e0fa632ab142b905))
* **CHIST-1:** first release ([bdab5f8](https://github.com/baloise/os3-copier/commit/bdab5f8c5590785286f011f8e6e5ae754bde153c))


controllers/suite_test.go
/*
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package controllers

import (
	"path/filepath"
	"testing"

	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
	"k8s.io/client-go/kubernetes/scheme"
	"k8s.io/client-go/rest"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/envtest"
	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
	"sigs.k8s.io/controller-runtime/pkg/log/zap"

	resourcebaloisechv1alpha1 "github.com/baloise/os3-copier/api/v1alpha1"
	// +kubebuilder:scaffold:imports
)

// These tests use Ginkgo (BDD-style Go testing framework). Refer to
// http://onsi.github.io/ginkgo/ to learn more about Ginkgo.

var cfg *rest.Config
var k8sClient client.Client
var testEnv *envtest.Environment

func TestAPIs(t *testing.T) {
	RegisterFailHandler(Fail)

	RunSpecsWithDefaultAndCustomReporters(t,
		"Controller Suite",
		[]Reporter{printer.NewlineReporter{}})
}

var _ = BeforeSuite(func(done Done) {
	logf.SetLogger(zap.LoggerTo(GinkgoWriter, true))

	By("bootstrapping test environment")
	testEnv = &envtest.Environment{
		CRDDirectoryPaths: []string{filepath.Join("..", "config", "crd", "bases")},
	}

	var err error
	cfg, err = testEnv.Start()
	Expect(err).ToNot(HaveOccurred())
	Expect(cfg).ToNot(BeNil())

	err = resourcebaloisechv1alpha1.AddToScheme(scheme.Scheme)
	Expect(err).NotTo(HaveOccurred())

	// +kubebuilder:scaffold:scheme

	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
	Expect(err).ToNot(HaveOccurred())
	Expect(k8sClient).ToNot(BeNil())

	close(done)
}, 60)

var _ = AfterSuite(func() {
	By("tearing down the test environment")
	err := testEnv.Stop()
	Expect(err).ToNot(HaveOccurred())
})


api/v1alpha1/groupversion_info.go
/*


Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package v1alpha1 contains API Schema definitions for the resource.baloise.ch v1alpha1 API group
// +kubebuilder:object:generate=true
// +groupName=resource.baloise.ch
package v1alpha1

import (
	"k8s.io/apimachinery/pkg/runtime/schema"
	"sigs.k8s.io/controller-runtime/pkg/scheme"
)

var (
	// GroupVersion is group version used to register these objects
	GroupVersion = schema.GroupVersion{Group: "resource.baloise.ch", Version: "v1alpha1"}

	// SchemeBuilder is used to add go types to the GroupVersionKind scheme
	SchemeBuilder = &scheme.Builder{GroupVersion: GroupVersion}

	// AddToScheme adds the types in this group-version to the given scheme.
	AddToScheme = SchemeBuilder.AddToScheme
)


api/v1alpha1/copyresource_types.go
/*


Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package v1alpha1

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.

// CopyResourceSpec defines the desired state of CopyResource
type CopyResourceSpec struct {
	// The Kind of the Resource you like to copy
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Enum=Secret;ConfigMap
	Kind string `json:"kind"`

	// The MetaName of the Resource found in metadata.name
	// +kubebuilder:validation:Required
	MetaName string `json:"metaName"`

	// The TargetNamespace the Resource should be copied to
	// +kubebuilder:validation:Required
	TargetNamespace string `json:"targetNamespace"`

	// The TargetName the Resource should be named in TargetNamespace
	// +kubebuilder:validation:Optional
	TargetName string `json:"targetName"`
}

// CopyResourceStatus defines the observed state of CopyResource
type CopyResourceStatus struct {
	ResourceVersion string `json:"resourceVersion"`
}

// +kubebuilder:object:root=true
// +kubebuilder:subresource:status

// CopyResource is the Schema for the copyresources API
type CopyResource struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   CopyResourceSpec   `json:"spec,omitempty"`
	Status CopyResourceStatus `json:"status,omitempty"`
}

// +kubebuilder:object:root=true

// CopyResourceList contains a list of CopyResource
type CopyResourceList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []CopyResource `json:"items"`
}

func init() {
	SchemeBuilder.Register(&CopyResource{}, &CopyResourceList{})
}


api/v1alpha1/zz_generated.deepcopy.go
// +build !ignore_autogenerated

/*


Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Code generated by controller-gen. DO NOT EDIT.

package v1alpha1

import (
	runtime "k8s.io/apimachinery/pkg/runtime"
)

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *CopyResource) DeepCopyInto(out *CopyResource) {
	*out = *in
	out.TypeMeta = in.TypeMeta
	in.ObjectMeta.DeepCopyInto(&out.ObjectMeta)
	out.Spec = in.Spec
	out.Status = in.Status
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new CopyResource.
func (in *CopyResource) DeepCopy() *CopyResource {
	if in == nil {
		return nil
	}
	out := new(CopyResource)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyObject is an autogenerated deepcopy function, copying the receiver, creating a new runtime.Object.
func (in *CopyResource) DeepCopyObject() runtime.Object {
	if c := in.DeepCopy(); c != nil {
		return c
	}
	return nil
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *CopyResourceList) DeepCopyInto(out *CopyResourceList) {
	*out = *in
	out.TypeMeta = in.TypeMeta
	in.ListMeta.DeepCopyInto(&out.ListMeta)
	if in.Items != nil {
		in, out := &in.Items, &out.Items
		*out = make([]CopyResource, len(*in))
		for i := range *in {
			(*in)[i].DeepCopyInto(&(*out)[i])
		}
	}
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new CopyResourceList.
func (in *CopyResourceList) DeepCopy() *CopyResourceList {
	if in == nil {
		return nil
	}
	out := new(CopyResourceList)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyObject is an autogenerated deepcopy function, copying the receiver, creating a new runtime.Object.
func (in *CopyResourceList) DeepCopyObject() runtime.Object {
	if c := in.DeepCopy(); c != nil {
		return c
	}
	return nil
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *CopyResourceSpec) DeepCopyInto(out *CopyResourceSpec) {
	*out = *in
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new CopyResourceSpec.
func (in *CopyResourceSpec) DeepCopy() *CopyResourceSpec {
	if in == nil {
		return nil
	}
	out := new(CopyResourceSpec)
	in.DeepCopyInto(out)
	return out
}

// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
func (in *CopyResourceStatus) DeepCopyInto(out *CopyResourceStatus) {
	*out = *in
}

// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new CopyResourceStatus.
func (in *CopyResourceStatus) DeepCopy() *CopyResourceStatus {
	if in == nil {
		return nil
	}
	out := new(CopyResourceStatus)
	in.DeepCopyInto(out)
	return out
}


controllers/copyresource_controller.go
/*


Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package controllers

import (
	"context"
	"fmt"
	v1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/apimachinery/pkg/types"

	"github.com/go-logr/logr"
	"k8s.io/apimachinery/pkg/runtime"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"

	resourcebaloisechv1alpha1 "github.com/baloise/os3-copier/api/v1alpha1"
	"github.com/jinzhu/copier"
)

// CopyResourceReconciler reconciles a CopyResource object
type CopyResourceReconciler struct {
	client.Client
	Log    logr.Logger
	Scheme *runtime.Scheme
}

type Object interface {
	metav1.Object
	runtime.Object
}

// +kubebuilder:rbac:groups=resource.baloise.ch,resources=copyresources,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=resource.baloise.ch,resources=copyresources/status,verbs=get;update;patch
// +kubebuilder:rbac:groups=resource.baloise.ch,resources=copyresources/finalizers,verbs=update
// +kubebuilder:rbac:groups=,resources=secrets,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=,resources=secrets/status,verbs=get;update;patch
// +kubebuilder:rbac:groups=,resources=secrets/finalizers,verbs=update
// +kubebuilder:rbac:groups=,resources=configmaps,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=,resources=configmaps/finalizers,verbs=update
// +kubebuilder:rbac:groups=,resources=configmaps/status,verbs=get;update;patch

func (r *CopyResourceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
	log := r.Log.WithValues("CopyResource", req.NamespacedName)

	copyResource := &resourcebaloisechv1alpha1.CopyResource{}
	err := r.Get(context.TODO(), req.NamespacedName, copyResource)
	if err != nil {
		if errors.IsNotFound(err) {
			log.Info("CopyResource not found. Ignoring since object must be deleted.")
			return ctrl.Result{}, nil
		}
		log.Error(err, "Failed to get CopyResource.", "namespacedName", req.NamespacedName)
		return ctrl.Result{}, nil
	}

	namespacedName := types.NamespacedName{
		Namespace: req.Namespace,
		Name:      copyResource.Spec.MetaName,
	}

	sourceResource, _ := StringToStruct(copyResource.Spec.Kind)
	err = r.Client.Get(context.TODO(), namespacedName, sourceResource)
	if err != nil && errors.IsNotFound(err) {
		log.Info("Source resource not found.", "namespacedName", namespacedName)
		return ctrl.Result{}, nil
	}
	if err != nil {
		log.Error(err, "Source resource error.", "namespacedName", namespacedName)
		return ctrl.Result{}, nil
	}

	targetResource, _ := StringToStruct(copyResource.Spec.Kind)
	targetResource, _ = cloneResource(copyResource.Spec.Kind, sourceResource, targetResource)
	targetResource.SetResourceVersion("")
	targetResource.SetUID("")
	targetResource.SetNamespace(copyResource.Spec.TargetNamespace)
	if copyResource.Spec.TargetName != "" {
		targetResource.SetName(copyResource.Spec.TargetName)
	} else {
		targetResource.SetName(copyResource.Namespace + "-" + copyResource.Name)
	}
	targetResource.SetOwnerReferences([]metav1.OwnerReference{buildOwnerReferenceToCopyResource(copyResource)})

	exists := isObjectExists(r, targetResource, log)

	if copyResource.Status.ResourceVersion == "" ||
		sourceResourceVersionHasChanged(copyResource.Spec.Kind, copyResource.Status.ResourceVersion, sourceResource) ||
		!exists {

		if !exists {
			err = r.Client.Create(context.TODO(), targetResource)
			if err != nil {
				log.Error(err, "Failed to create resource.", "name", targetResource.GetName(), "namespace ", targetResource.GetNamespace())
				return ctrl.Result{}, nil
			}
			log.Info("Successfully created.", "name", targetResource.GetName(), "namespace ", targetResource.GetNamespace())
		} else {
			err = r.Client.Update(context.TODO(), targetResource)
			if err != nil {
				log.Error(err, "Failed to update.", "name", targetResource.GetName(), "namespace ", targetResource.GetNamespace())
				return ctrl.Result{}, nil
			}
			log.Info("Successfully update.", "name", targetResource.GetName(), "namespace ", targetResource.GetNamespace())
		}

		copyResource.Status.ResourceVersion = getResourceVersion(copyResource.Spec.Kind, sourceResource)
		err := r.Status().Update(context.TODO(), copyResource)
		if err != nil {
			log.Error(err, "Failed to update CopyResource status.", "resourceVersion", copyResource.Status.ResourceVersion)
			return ctrl.Result{}, nil
		}
	}

	return ctrl.Result{}, nil
}

func (r *CopyResourceReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(&resourcebaloisechv1alpha1.CopyResource{}).
		Complete(r)
}

func isObjectExists(r *CopyResourceReconciler, targetResource Object, log logr.Logger) bool {
	targetNamespacedName := types.NamespacedName{
		Namespace: targetResource.GetNamespace(),
		Name:      targetResource.GetName(),
	}
	// Use an unstructured type to avoid cache reader
	u := &unstructured.Unstructured{}
	u.SetGroupVersionKind(schema.GroupVersionKind{
		Group:   "",
		Kind:    targetResource.GetObjectKind().GroupVersionKind().Kind,
		Version: "v1",
	})
	err := r.Client.Get(context.TODO(), targetNamespacedName, u)
	if err != nil {
		if errors.IsNotFound(err) {
			log.Info("Not found " + targetNamespacedName.Namespace + "/" + targetNamespacedName.Name)
		}
		return false
	}
	return true
}

func StringToStruct(kind string) (Object, error) {
	switch kind {
	case "Secret":
		return &v1.Secret{}, nil
	case "ConfigMap":
		return &v1.ConfigMap{}, nil
	default:
		return nil, fmt.Errorf("%s is not a known resource kind", kind)
	}
}

func cloneResource(kind string, source Object, target Object) (Object, error) {
	switch kind {
	case "Secret":
		sourceSecret := source.(*v1.Secret)
		copier.Copy(target.(*v1.Secret), sourceSecret)
		return target, nil
	case "ConfigMap":
		sourceConfigMap := source.(*v1.ConfigMap)
		copier.Copy(target.(*v1.ConfigMap), sourceConfigMap)
		return target, nil
	default:
		return nil, fmt.Errorf("%s is not a known resource kind", kind)
	}
}

func buildOwnerReferenceToCopyResource(copyResource *resourcebaloisechv1alpha1.CopyResource) metav1.OwnerReference {
	return metav1.OwnerReference{
		APIVersion: copyResource.APIVersion,
		Kind:       copyResource.Kind,
		Name:       copyResource.GetName(),
		UID:        copyResource.GetUID(),
		// If true, this reference points to the managing controller.
		Controller: BoolPointer(true),
		// Don't block owner deletion (CopyResource) if this resource still exists
		BlockOwnerDeletion: BoolPointer(false),
	}
}

func sourceResourceVersionHasChanged(kind string, copyResourceVersion string, source Object) bool {
	sourceResourceVersion := getResourceVersion(kind, source)
	return sourceResourceVersion != copyResourceVersion
}

func getResourceVersion(kind string, resource Object) string {
	switch kind {
	case "Secret":
		return resource.(*v1.Secret).ResourceVersion
	case "ConfigMap":
		return resource.(*v1.ConfigMap).ResourceVersion
	default:
		return ""
	}
}

func BoolPointer(b bool) *bool {
	return &b
}


main.go
/*
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"fmt"
	"go.uber.org/zap/zapcore"
	"os"
	"strconv"
	"time"

	"k8s.io/apimachinery/pkg/runtime"
	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/healthz"
	"sigs.k8s.io/controller-runtime/pkg/log/zap"

	resourcebaloisechv1alpha1 "github.com/baloise/os3-copier/api/v1alpha1"
	"github.com/baloise/os3-copier/controllers"
	// +kubebuilder:scaffold:imports
)

const (
	WatchNamespaceEnvName = "WATCH_NAMESPACE"
	SyncPeriodEnvName     = "SYNC_PERIOD"
)

var (
	scheme   = runtime.NewScheme()
	setupLog = ctrl.Log.WithName("setup")
)

func init() {
	utilruntime.Must(clientgoscheme.AddToScheme(scheme))

	utilruntime.Must(resourcebaloisechv1alpha1.AddToScheme(scheme))
	// +kubebuilder:scaffold:scheme
}

func main() {
	var metricsAddr string
	var healtAddr string
	var enableLeaderElection bool
	var devModeEnabled bool
	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
	flag.StringVar(&healtAddr, "probe-addr", ":8081", "The address the health check endpoint binds to.")
	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
		"Enable leader election for controller manager. "+
			"Enabling this will ensure there is only one active controller manager.")
	flag.BoolVar(&devModeEnabled, "dev-mode-enabled", false,
		"Enable dev mode to see DEBUG logs and stack traces. ")
	flag.Parse()

	var stacktraceLevel zapcore.LevelEnabler
	var logLevel zapcore.LevelEnabler
	if devModeEnabled {
		stacktraceLevel = zapcore.WarnLevel
		logLevel = zapcore.DebugLevel
	} else {
		stacktraceLevel = zapcore.FatalLevel
		logLevel = zapcore.InfoLevel
	}
	ctrl.SetLogger(zap.New(zap.UseDevMode(true), zap.StacktraceLevel(stacktraceLevel), zap.Level(logLevel)))

	watchNamespace, err := getEnvVar(WatchNamespaceEnvName)
	if err != nil {
		setupLog.Error(err, "Unable to get WatchNamespace, "+
			"please set environment variable "+WatchNamespaceEnvName)
		os.Exit(1)
	}
	setupLog.Info(WatchNamespaceEnvName + " set, using " + watchNamespace + "as namespace")

	syncPeriod := getSyncPeriod()
	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
		Scheme:                 scheme,
		MetricsBindAddress:     metricsAddr,
		LivenessEndpointName:   "/healthz",
		ReadinessEndpointName:  "/readyz",
		HealthProbeBindAddress: healtAddr,
		Port:                   9443,
		LeaderElection:         enableLeaderElection,
		LeaderElectionID:       "3dacd622.baloise.ch",
		SyncPeriod:             &syncPeriod,
		Namespace:              watchNamespace,
	})
	if err != nil {
		setupLog.Error(err, "unable to start manager")
		os.Exit(1)
	}

	if err = (&controllers.CopyResourceReconciler{
		Client: mgr.GetClient(),
		Log:    ctrl.Log.WithName("controllers").WithName("CopyResource"),
		Scheme: mgr.GetScheme(),
	}).SetupWithManager(mgr); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "CopyResource")
		os.Exit(1)
	}
	// +kubebuilder:scaffold:builder

	if err := mgr.AddHealthzCheck("health", healthz.Ping); err != nil {
		setupLog.Error(err, "unable to set up health check")
		os.Exit(1)
	}
	if err := mgr.AddReadyzCheck("check", healthz.Ping); err != nil {
		setupLog.Error(err, "unable to set up ready check")
		os.Exit(1)
	}

	setupLog.Info("starting manager for os3-copier.")
	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
		setupLog.Error(err, "problem running manager")
		os.Exit(1)
	}
}

func getEnvVar(name string) (string, error) {
	ns, found := os.LookupEnv(name)
	if !found {
		return "", fmt.Errorf("%s must be set", name)
	}
	return ns, nil
}

func getSyncPeriod() time.Duration {
	syncPeriodInSeconds, err := getEnvVar(SyncPeriodEnvName)
	syncPeriodInSecondsInt, err := strconv.ParseInt(syncPeriodInSeconds, 10, 64)
	if err != nil {
		setupLog.Info(SyncPeriodEnvName + " not set, using 300s as default syncPeriod")
		return time.Duration(300) * time.Second
	}
	setupLog.Info(SyncPeriodEnvName + " set, using " + syncPeriodInSeconds + "s as syncPeriod")
	return time.Duration(syncPeriodInSecondsInt) * time.Second
}
